### PR TITLE
Experimental BuildKit support

### PR DESCRIPTION
**Pull request by tonistiigi**
This PR adds experimental support for using [moby/buildkit](https://github.com/moby/buildkit) as a backend for `docker build`. #32925 #34227

The support is only enabled in the experimental daemon, and user needs to opt-in in docker/cli to start using it.

```
# in daemon
dockerd -D --experimental

# in Docker CLI
export DOCKER_BUILDKIT=1
docker build .
```

Docker CLI branch with BuildKit support enabled is in https://github.com/tonistiigi/docker-cli/tree/experimental-buildkit

In the API, `version` field allows switching between different build backends.

This enables lots of new capabilities, including parallel build steps and requests, skipping unused stages, efficient incremental builds, build context file detection, remote cache support, graceful cancellation, build cache issues, etc. It also enables using experimental Dockerfile features without requiring to update Moby by using the BuildKit frontend images. 

The technical side of this integration is a temporary (and somewhat limited) solution as BuildKit is based on containerd storage and that has not been integrated into Moby yet. The current solution provides adapters to some components to provide compatibility between things like containerd snapshots and Moby layerstore/storage drivers or image store. When containerd storage support lands in Moby, all these adapters should be removed entirely, and BuildKit can use containerd directly. This temporary solution allows us to get feedback before we feel comfortable to move it out of experimental.

The tests will not currently use BuildKit in the PR as it requires opt-in and a new CLI binary. We have run the tests manually. Some of them need to be reworked for output change or a side-effect removal. We will continue addressing some of the test results that we are still investigating.

Known issues (some of them will not make it into 18.06 release):
- no automatic GC (manual `system df` and `system prune` work)
- ~pulling schema1 images during build~
- ~automatic path detection in dockerfile doesn't detect unused symlinks~
- resource limitation and --net options: execution completely bypasses the daemon, net is host.
- windows support (depending on containerd runtime integration)
- `--pull` (currently existing images in image store always take precedence)
- USER doesn't set additional groups
- `--iidfile` is not enabled in CLI
- No way to export remote build cache. Build cache exported with standalone BuildKit can be imported with `--cache-from`.

@tiborvass @AkihiroSuda

depends on #36895

[asciicast](https://asciinema.org/a/B1btlPB1N4tGiqVvVfZjWeuoQ)

### CODE REVIEW

File: builder/builder-next/adapters/snapshot/snapshot.go
Comment (AkihiroSuda) [inline builder/builder-next/adapters/snapshot/snapshot.go]: Can we split this to a separate package that wraps Moby graphdriver as a containerd snapshotter?

cc @dmcgowan

File: builder/builder-next/exporter/writer.go
Comment (AkihiroSuda) [inline builder/builder-next/exporter/writer.go]: Can we dedupe writer.go with https://github.com/moby/buildkit/blob/master/exporter/containerimage/writer.go via some util pkg?

File: api/types/client.go
Comment (AkihiroSuda) [inline api/types/client.go]: Nit: can we just use strings rather than numbers? e.g. `legacy` and `buildkit`

File: api/types/client.go
Comment (AkihiroSuda) [inline api/types/client.go]: Could you add godoc?

File: api/types/client.go
Comment (tiborvass) [reply api/types/client.go]: @AkihiroSuda I don't like the word `legacy` (which is what it was initially). With all the respect I have for buildkit, with time everything becomes legacy and eventually you'll have to differentiate between `legacy legacy` and `legacy`. This problem is solved with versions, so I believe it's a better tradeoff this way. Happy to hear your concerns though.

File: builder/builder-next/adapters/snapshot/snapshot.go
Comment (tonistiigi) [reply builder/builder-next/adapters/snapshot/snapshot.go]: Currently, the issue is that containerd mounts are stateless while some graphdrives can't produce stateless mounts. @dmcgowan Has a plan for that for Moby integration that is somewhat similar to the `buildkit/snapshot.Mountable` abstraction.

File: builder/builder-next/exporter/writer.go
Comment (tonistiigi) [reply builder/builder-next/exporter/writer.go]: The reason I couldn't use it directly was that buildkit uses the `blobsum-diffid` pairs here while Moby doesn't yet have an understanding of managing layer blobs. But with a careful refactor, some parts of it can definitely be shared.

**Review by vdemeester — COMMENTED** [review]
On high level, definitely Design LGTM :heart_eyes_cat: 

In this PR it seems it's not possible to choose a different frontend yet right ? if yes, is it something we plan to allow before going out of experimental ?
I see the `gateway.v0` frontend is there so it seems it's already possible to use custom images as frontend (which is awesome :heartbeat:) but I'm not sure how to specify it from the API standpoint. 

Also, should we open a PR in `docker/cli` to discuss the `cli` design in parallel ; and also easily have a docker binary built to make manual testing easier (without having to clone and build the cli)

**Comment by tonistiigi** [comment]
> In this PR it seems it's not possible to choose a different frontend yet right ? if yes, is it something we plan to allow before going out of experimental ? I see the gateway.v0 frontend is there so it seems it's already possible to use custom images as frontend (which is awesome 💓) but I'm not sure how to specify it from the API standpoint.

Custom Dockerfile implementations can be used with a Dockerfile directive https://github.com/moby/buildkit/pull/384 but generic frontends (or raw LLB) is not currently enabled in HTTP API. As this would be a new feature, it probably makes sense to do discuss/PR that separately. It may definitely happen in experimental is maintainers agree.

> Also, should we open a PR in docker/cli to discuss the cli design in parallel ; and also easily have a docker binary built to make manual testing easier (without having to clone and build the cli)

There's still some cleanup stuff that needs to happen there. You can expect a PR early next week.

**Comment by cpuguy83** [comment]
Design LGTM

File: vendor.conf
Comment (thaJeztah) [inline vendor.conf]: Is this something we can upstream? I didn't see a pull request; https://github.com/hashicorp/go-immutable-radix/pulls (looks like its this change; https://github.com/tonistiigi/go-immutable-radix/commit/826af9ccf0feeee615d546d69b11f8e98da8c8f1)

**Review by thaJeztah — CHANGES_REQUESTED** [review]

File: vendor.conf
Comment (tonistiigi) [reply vendor.conf]: Not sure if they are interested and the project isn't very active. I can make a proper fork if needed (although we use this project for swarmkit as well). There is no functionality divergence, just an extra method that is only used by buildkit.

Comment (thaJeztah) [reply vendor.conf]: We could try if they accept the PR, then we'd be set (unless we think we'll be adding more features and a fork would help)

Comment (tonistiigi) [reply vendor.conf]: Ok, I'll make sure to do that and replace if it gets merged

File: vendor.conf
Comment (tonistiigi) [reply vendor.conf]: removed

**Comment by AkihiroSuda** [comment]
Non-BuildKit mode seems broken?

```
suda@suda-ws01:/tmp/fuga$ docker version
Client:
 Version:      18.06.0-dev
 API version:  1.38
 Go version:   go1.10.2
 Git commit:   99619587
 Built:        Sat Jun  9 05:04:49 2018
 OS/Arch:      linux/amd64
 Experimental: true
 Orchestrator: swarm

Server:
 Engine:
  Version:      dev
  API version:  1.38 (minimum version 1.12)
  Go version:   go1.10.2
  Git commit:   a58f4062c
  Built:        Sat Jun  9 05:26:19 2018
  OS/Arch:      linux/amd64
  Experimental: true

suda@suda-ws01:/tmp/fuga$ cat Dockerfile 
FROM alpine
RUN apk update

suda@suda-ws01:/tmp/fuga$ docker build -t foo .
Sending build context to Docker daemon  2.048kB
Failed to parse aux message: json: cannot unmarshal string into Go value of type types.BuildResultFailed to parse aux message: json: cannot unmarshal string into Go value of type types.BuildResultFailed to parse aux message: json: cannot unmarshal string into Go value of type types.BuildResultFailed to parse aux message: json: cannot unmarshal string into Go value of type types.BuildResultFailed to parse aux message: json: cannot unmarshal string into Go value of type types.BuildResultFailed to parse aux message: json: cannot unmarshal string into Go value of type types.BuildResultFailed to parse aux message: json: cannot unmarshal string into Go value of type types.BuildResultFailed to parse aux message: json: cannot unmarshal string into Go value of type types.BuildResultFailed to parse aux message: json: cannot unmarshal string into Go value of type types.BuildResultFailed to parse aux message: json: cannot unmarshal string into Go value of type types.BuildResultFailed to parse aux message: json: cannot unmarshal string into Go value of type types.BuildResultFailed to parse aux message: json: cannot unmarshal string into Go value of type types.BuildResultFailed to parse aux message: json: cannot unmarshal string into Go value of type types.BuildResultFailed to parse aux message: json: cannot unmarshal string into Go value of type types.BuildResultFailed to parse aux message: json: cannot unmarshal string into Go value of type types.BuildResult^C
```

**Comment by tonistiigi** [comment]
@tiborvass Issue with latest CLI branch. previous branch (7979fb6ab) doesn't seem to have problems.

**Review by tiborvass — APPROVED** [review]
LGTM

**Comment by tiborvass** [comment]
It's really unfortunate to have all these unrelated CI issues.

**Review by AkihiroSuda — APPROVED** [review]
LGTM if the regression is fixed now.

**Comment by tiborvass** [comment]
@AkihiroSuda yes, regression was on docker CLI.

**Review by vdemeester — APPROVED** [review]
LGTM :lion:

**Comment by tiborvass** [comment]
CI for powerpc and s390x are known to have issues /cc @andrewhsu @seemethere.

**Comment by jhowardmsft** [comment]
This has broken master on Windows:

```
PS C:\> dockerd
INFO[2018-06-13T09:04:23.208583500-07:00] Windows default isolation mode: process
INFO[2018-06-13T09:04:23.294671100-07:00] Loading containers: start.
INFO[2018-06-13T09:04:23.298646300-07:00] Restoring existing overlay networks from HNS into docker
INFO[2018-06-13T09:04:23.399254200-07:00] Loading containers: done.
INFO[2018-06-13T09:04:23.402261600-07:00] Docker daemon                                 commit=692df4699c graphdriver(s)=windowsfilter version=0.0.0-dev
INFO[2018-06-13T09:04:23.405258500-07:00] Daemon has completed initialization
failed to find git binary: exec: "git": executable file not found in %PATH%
PS C:\>
```

**Comment by tiborvass** [comment]
@jhowardmsft Thanks. Sorry about that. Windows CI was green :(

**Comment by jhowardmsft** [comment]
@tiborvass Yes, the CI servers have (and need) git installed :/

**Pull request merged**
